grant-project/zcash-grant-system
# Blockchain watcher bootstrap

No issue associated with this.

### What This Does

Due to the stateless nature of the watcher, restarts cause is to lose the contribution addresses we're tracking, and start us at the current block. This changes it so that the backend will ping the watcher with a bootstrapping payload that gets the watcher watching any outstanding contributions, from the last one to be validated.

This implementation is still a little bit janky, and often leads to double-bootstraps. Looking for any feedback on how to make it better.

### Steps to Test

1. Start your blockchain watcher, confirm it doesn't start scanning blocks.
2. Start your backend and load a page, confirm it starts scanning blocks.
3. Do the same in reverse, start the backend followed by the watcher. Confirm the watcher starts watching blocks.
4. Make a contribution and get it to confirm (wait for testnet, run `generate` for regtest.)
5. After it confirms, kill your blockchain watcher and start it again. Confirm it starts scanning & watching from the block where the transaction was confirmed (minus minimum confirmation env var.)
